Improve request evaluator cycle diagnostics

beccadax opened the pull request:
This PR:

* Refactors `DiagnosticEngine`'s declaration-pretty-printing facility to allow it to be used from `extractNearestSourceLoc()`, improving cycle diagnostics that were previously emitted at `<invalid>:0`.
* Modifies the `-debug-cycles` output to indicate which request in the dumped stack was repeated.

~~I haven't figured out how to test these changes yet, which is why this is still currently a WIP.~~

beccadax committed:
[NFC] Indicate offending request in -dump-cycles

beccadax commented:
@swift-ci please smoke test

beccadax commented on test/CircularReferences/Inputs/custom-modules/CircularLibrary.swift:
```diff
@@ -0,0 +1,17 @@
+// swift-interface-format-version: 1.0
+// swift-module-flags: -module-name CircularLibrary
+
+//import Swift
```
Nit: I should delete these lines.

slavapestov commented:
Nice!

AnthonyLatsis commented:
@beccadax There's a [PR](https://github.com/apple/swift/pull/35849) with little traction that restores the way `-debug-cycles` output used to look like, in case you'd like to consider incorporating those changes.

beccadax commented:
@swift-ci please test

beccadax commented:
The previous test case was very fragile and is, in itself, a bug we probably ought to fix. I've filed that bug as SR-14324 and added a command-line argument to the frontend which artificially creates a request cycle involving arbitrary types that we can diagnose.

beccadax commented:
@swift-ci please test Linux platform

beccadax commented:
@swift-ci please smoke test

beccadax commented:
@swift-ci please smoke test

beccadax committed:
Use pretty-printed decls in cycle diagnostics

This commit makes the pretty-printing behavior of `DiagnosticEngine` easily accessible via a new `Decl::getLocByPrintingIfNeeded()` call and uses that call in various places related to `extractNearestSourceLoc()`, ultimately improving diagnostics for cycles involving serialized declarations.

Note that the test case added here should probably not cause a circularity error; SR-14324 tracks this bug.

beccadax committed:
[NFC] Add pure test of pretty-printed cycle errors

This commit implements a facility to artificially cause a request cycle involving various types so we can test how circularity errors print without having to carefully craft test cases where they happen.

beccadax commented:
@swift-ci please smoke test

beccadax commented:
@swift-ci please smoke test

beccadax commented:
@swift-ci please smoke test

beccadax commented:
@swift-ci please smoke test macOS platform
